src/transformStateWithClones.js

```diff
@@ -6,8 +6,65 @@
  *
  * @return {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const hystoryArray = [];
+  const stateCopy = { ...state };
+
+  for (const action of actions) {
+    const newCopy = Object.assign({}, hystoryArray[hystoryArray.length - 1]);
+
+    switch (action.type) {
+      case 'addProperties':
+        if (hystoryArray.length >= 1) {
+          hystoryArray.push(addProperties(newCopy, action.extraData));
+          break;
+        } else {
+          hystoryArray.push(addProperties(stateCopy, action.extraData));
+          break;
+        }
+
+      case 'removeProperties':
+        if (hystoryArray.length >= 1) {
+          hystoryArray.push(removeProperties(newCopy, action.keysToRemove));
```

Comment:
Instead of pushing the result inside each case, consider modifying the state copy and pushing it to the array at the end of the loop cycle, outside of the `switch` block. This will reduce redundancy and improve code readability.